trillium/src/conn.rs

```diff
@@ -547,6 +547,25 @@ impl Conn {
         })
     }
 
+    /// Attempt to get a reference to the trait object as a specific transport type T.
+    ///
+    /// If the type does not match the original transport type, this will return `None`.
+    pub fn downcast_transport_ref<T: Transport>(&self) -> Option<&T> {
```

Comment:
I'm on the fence about introducing these to the trillium::Conn instead of requiring folks who need this advanced interface drop down to trillium_http::Conn. On one hand, it would be nice to do away with `inner` and `inner_mut`, on the other, that might be the right way to indicate to someone that what they're doing is advanced, and eventually gate `inner` and `inner_mut` on a trillium feature flag with the semantics of `#[cfg(feature = "i-acknowledge-that-i-am-leaving-the-well-trod-path")]`

Reply:
I'm also on the fence about the interface on Conn. Maybe this usage requires `trillium_conn.inner().transport().as_any().downcast_ref()` / `trillium_conn.inner_mut().transport_mut().as_any_mut().downcast_mut()` to make it harder to stumble into?

Reply:
I really like the currently documented premise that nothing should require dropping down to the `trillium_http::Conn`. If you want the top-level methods removed I can do that, but I think going from "if you need to do `.inner()` that's always a bug" to "do `.inner()` for various less-common methods" seems less ideal.

Reply:
> I'm also on the fence about the interface on `trillium_http::Conn<BoxedTransport>`. Maybe this usage requires `trillium_conn.inner().transport().as_any().downcast_ref()` / `trillium_conn.inner_mut().transport_mut().as_any_mut().downcast_mut()` to make it harder to stumble into?

Thinking about this: this would still require exposing some new methods, namely `transport()`/`transport_mut()` and `as_any()`/`as_any_mut()`. However, if you'd prefer this, I can implement it.